unpublished/build-system-abstraction-njs.rst

```diff
@@ -0,0 +1,708 @@
+configuration -- so pretty much all we really need from a build system
+is that it have some way to spit out standard-compliant wheels and
+sdists.
+
+We therefore propose a new, relatively minimal interface for
+installation tools like ``pip`` to interact with package source trees
+and source distributions.
+
+
+=======================
+ Terminology and goals
+=======================
+
+A *source tree* is something like a VCS checkout. We need a standard
+interface for installing from this format, to support usages like
+``pip install some-directory/``.
+
+A *source distribution* is a static snapshot representing a particular
+release of some source code, like ``lxml-3.4.4.zip``. Source
+distributions serve many purposes: they form an archival record of
+releases, they provide a stupid-simple de facto standard for tools
+that want to ingest and process large corpora of code, possibly
+written in many languages (e.g. code search), they act as the input to
+downstream packaging systems like Debian/Fedora/Conda/..., and so
+forth. In the Python ecosystem they additionally have a particularly
+important role to play, because packaging tools like ``pip`` are able
+to use source distributions to fulfill binary dependencies, e.g. if
+there is a distribution ``foo.whl`` which declares a dependency on
+``bar``, then we need to support the case where ``pip install bar`` or
+``pip install foo`` automatically locates the sdist for ``bar``,
+downloads it, builds it, and installs the resulting package.
+
+Source distributions are also known as *sdists* for short.
+
+A *build frontend* is a tool that users might run that takes arbitrary
+source trees or source distributions and builds wheels from them. The
+actual building is done by each source tree's *build backend*. In a
+command like ``pip wheel some-directory/``, pip is acting as a build
+frontend.
+
+An *integration frontend* is a tool that users might run that takes a
+set of package requirements (e.g. a requirements.txt file) and
+attempts to update a working environment to satisfy those
+requirements. This may require locating, building, and installing a
+combination of wheels and sdists. In a command like ``pip install
+lxml==2.4.0``, pip is acting as an integration frontend.
+
+
+==============
+ Source trees
+==============
+
+We retroactively declare the legacy source tree format involving
+``setup.py`` to be "version 0". We don't try to specify it further;
+its de facto specification is encoded in the source code and
+documentation of ``distutils``, ``setuptools``, ``pip``, and other
+tools.
+
+A "version 1" (or greater) source tree is any directory which contains
+a file named ``pypackage.json``. (If a tree contains both
+``pypackage.json`` and ``setup.py`` then it is a version 1+ source
+tree and the ``setup.py`` is ignored; this allows packages to include
+a ``setup.py`` for compatibility with old build frontends, while using
+the new system with new build frontends.)
+
```

Comment:
Unlike Robert's proposal, this one explicitly *does* get into the business of defining a sdist format (via defining the "source tree" format). And it does so in a way that ignores the possibility of static (= introspectable without needing to involve the build system) metadata. And what does it gain us to be any more formal than Robert's PEP? ("The build interface is defined in terms of a source tree that contains a single config file and "other stuff" as needed by the build system"). We don't provide any benefits to tools that *don't* need to do builds (consider twine, which in uploading a sdist may benefit from being able to extract the metadata and add it to the PyPI project metadata - but which doesn't want to go near the build system).

I still think I'd like to see a formal sdist format at some point, simply so that we have a reference for how tools can introspect any file they find on PyPI, but the discussion on Robert's PEP has convinced me that build system abstraction proposals are not the right place to get into that.